File: leetcode/merge-intervals.java
// https://leetcode.com/problems/merge-intervals/

import java.util.ArrayList;
import java.util.Arrays;

class Solution {
    public int[][] merge(int[][] intervals) {
        ArrayList<int[]> intervalList = new ArrayList<int[]>();
        
        Arrays.sort(intervals, (one, two) -> Integer.compare(one[0], two[0]));
        
        int index = 0;
        while(index < intervals.length){
            int[] interval = new int[]{intervals[index][0], intervals[index][1]};
            index++;
            while(index < intervals.length && intervals[index][0] >= interval[0] && intervals[index][0] <= interval[1]){
                interval[1] = Math.max(interval[1], intervals[index++][1]);
            }
            intervalList.add(interval);
        }

        return intervalList.toArray(new int[intervalList.size()][2]);
    }
}

File: leetcode/merge_intervals.py
# https://leetcode.com/problems/merge-intervals/description/


class Solution:
    def merge(self, intervals: list[list[int]]) -> list[list[int]]:
        intervals.sort(key=lambda a: a[0])

        updated_intervals: list[list[int]] = []
        curr_interval: list[int] = intervals[0]

        for i in range(1, len(intervals)):
            if intervals[i][0] <= curr_interval[1]:
                curr_interval[1] = max(curr_interval[1], intervals[i][1])
            else:
                updated_intervals.append(curr_interval)
                curr_interval = intervals[i]

        updated_intervals.append(curr_interval)

        return updated_intervals
